bin/external_lib_deps.ml

```diff
@@ -1,28 +1,217 @@
+            let dir = dune_file.dir in
+            match stanza with
+            | Dune_file.Executables exes ->
+              Memo.return [ (exes.buildable.libraries, dir) ]
+            | Dune_file.Library lib ->
+              Memo.return [ (lib.buildable.libraries, dir) ]
+            | _ -> Memo.return [])
+        >>| List.concat)
+    >>| List.flatten
+  in
+  lib_deps
+
+let libs_to_lib_map libs =
+  List.fold_left
+    ~f:(fun acc_map (name, kind) ->
+      Lib_name.Map.update acc_map name ~f:(fun n ->
+          match n with
+          | Some k -> Some (Kind.merge k kind)
+          | None -> Some kind))
+    ~init:Lib_name.Map.empty libs
+
+let libs_dir_to_map libs_dir =
+  List.fold_left
+    ~f:(fun acc_map (dir, libs) ->
+      match Dir_map.find acc_map dir with
+      | None -> Dir_map.add_exn acc_map dir (libs_to_lib_map libs)
+      | Some libs_map ->
+        Dir_map.set acc_map dir
+          (Lib_name.Map.union libs_map (libs_to_lib_map libs) ~f:(fun _ k1 k2 ->
+               Some (Kind.merge k1 k2))))
+    ~init:Dir_map.empty libs_dir
+
+let libs_sort libs =
+  List.sort_uniq
+    ~compare:(fun (name1, _) (name2, _) ->
+      String.compare (Lib_name.to_string name1) (Lib_name.to_string name2))
+    libs
+
+let run ~context_name ~external_resolved_libs ~by_dir ~sexp ~only_missing =
+  if only_missing then (
+    if by_dir || sexp then
+      User_error.raise
+        [ Pp.textf "--only-missing cannot be used with --by_dir or --sexp" ];
+    let missing =
+      external_resolved_libs |> List.map ~f:snd |> List.concat
+      |> List.filter_map ~f:(fun ((result, kind_db), kind, name) ->
+             match (result, kind_db) with
+             | None, Lib.Kind_db.Installed_libs -> Some (name, kind)
+             | _ -> None)
+      |> libs_sort
+    in
+    if List.for_all ~f:(fun (_, kind) -> kind = Kind.Optional) missing then
+      User_message.prerr
+        (User_error.make
+           [ Pp.textf "The following libraries are missing in the %s context:"
+               context_name
+           ; pp_external_libs missing
+           ])
+    else
+      User_message.prerr
+        (User_error.make
+           [ Pp.textf "The following libraries are missing in the %s context:"
+               context_name
+           ; pp_external_libs missing
+           ]
+           ~hints:
+             [ Dune_engine.Utils.pp_command_hint
+               @@ (missing
+                  |> List.filter_map ~f:(fun (name, kind) ->
+                         match (kind : Kind.t) with
+                         | Optional -> None
+                         | Required -> Some (Lib_name.package_name name))
+                  |> List.map ~f:Package.Name.to_string
+                  |> String.concat ~sep:" ")
+             ]))
+  else if sexp then (
+    if not by_dir then User_error.raise [ Pp.textf "--sexp requires --by-dir" ];
+    let sexp =
+      external_resolved_libs
+      |> List.map ~f:(fun (dir, libs) ->
```

Comment:
This is duplicated with the branch below. I would recommend defining an appropriate data structure, build it in both cases, and print it differently (using to_dyn in one case)